### DISCUSSION TITLE
Are there any plans to introduce @Nullable/@NotNull annotations in the Ghidra source tree?

### DISCUSSION THREAD

**Discussion by agatti** [NationalSecurityAgency/ghidra#4723] [Ideas]
That would allow static analysis in [Eclipse](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Ftasks%2Ftask-using_null_annotations.htm) and other IDEs, and through both gradle and CI pipelines (see [SpotBugs](https://spotbugs.github.io), [PMD](https://pmd.github.io), [Qodana](https://www.jetbrains.com/qodana/jvm/), and so on).

Since all of this is annotation-driven, you don't have to add annotations to each and every method right away, but you can introduce them with new code first and then go on from there.  I understand that deprecating all the old code that doesn't use `Optional<T>` is not really viable, so annotations are the next best thing IMHO.

As an added bonus this would help the two or three of us who use non-Java JVM languages for writing extensions, where nullable and non-nullable variables are two separate things :)

**Comment by ryanmkurtz** [comment]
To answer your initial question, no, we do not have plans to introduce this.  I could defintely see us using it if it was built into the language or standardized in some way though.

**Comment by agatti** [comment]
Well, there's [JSR-305](https://jcp.org/en/jsr/detail?id=305) as far as standardization goes.  With Java 8 they decided to go with `Optional<T>` rather than use annotations, but in Ghidra's case that would mean rewriting (and deprecating) a fair amount of old code.

I personally use SpotBugs' annotations set, which aren't in the `javax.annotations` package, and configure the IDE to use those `@Nullable`, `@NotNull`, etc. annotations instead.  My build scripts also use the SpotBugs Gradle plugin to perform checks when needed.

However, if using an external analyzer isn't desirable, you can always create two empty annotations with the two names mentioned earlier and have Eclipse use them to check for nullability (see the first link in the earlier comment).
